src/mqtt/mqtt.cpp

```diff
@@ -231,190 +271,235 @@ void MQTT::on_connect(int connackCode)
+	if (!writeOpIsPending) {
 		return;
 	}
 
-	const auto socket = mosquittopp::socket();
+	auto result = m_mosquitto.client()->loopWrite();
+	MqttLib::instance().checkMosquittoResultAndDoDebugPrints(result, "loopWrite()");
+}
 
-	m_fdnRead = std::make_unique<FileDescriptorNotifier>(socket, FileDescriptorNotifier::NotificationType::Read);
-	m_fdnWrite = std::make_unique<FileDescriptorNotifier>(socket, FileDescriptorNotifier::NotificationType::Write);
+void MqttClient::onMiscTaskRequested()
+{
+	auto result = m_mosquitto.client()->loopMisc();
+	MqttLib::instance().checkMosquittoResultAndDoDebugPrints(result, "loopMisc()");
+}
+
+void MqttClient::EventLoopHook::init(const std::chrono::milliseconds miscTaskInterval, MqttClient *parent)
+{
+	spdlog::debug("MqttClient::EventLoopHook::init()");
+	assert(parent != nullptr);
 
-	m_fdnRead->triggered.connect([this]() { onFileDescriptorNotifierTriggeredRead(); } );
-	m_fdnWrite->triggered.connect([this]() { onFileDescriptorNotifierTriggeredWrite(); } );
+	this->parent = parent;
 
-	m_miscTaskTimer = std::make_unique<Timer>();
-	m_miscTaskTimer->interval.set(miscTaskTimerInterval);
-	m_miscTaskTimer->timeout.connect([this]() { onMiscTaskTimerTriggered(); } );
-	m_miscTaskTimer->running.set(true);
+	miscTaskTimer = std::make_unique<Timer>();
+	miscTaskTimer->interval.set(miscTaskInterval);
+	miscTaskTimer->running.set(false);
 }
 
-void MQTT::unhookFromEventLoop()
+void MqttClient::EventLoopHook::engage(const int socket)
 {
-	spdlog::debug("MQTT::unhookFromEventLoop()");
+	spdlog::debug("MqttClient::EventLoopHook::engage()");
 
-	if (!(m_fdnRead && m_fdnWrite && m_miscTaskTimer)) {
-		spdlog::error("MQTT::hookToEventLoop() - Already unhooked from event loop.");
+	if (!isSetup()) {
+		spdlog::error("MqttClient::EventLoopHook::engage() - EventLoopHook is not initialized. Call MqttClient::EventLoopHook::init() first.");
 		return;
 	}
 
-	m_fdnRead->triggered.disconnectAll();
-	m_fdnWrite->triggered.disconnectAll();
-	m_miscTaskTimer->timeout.disconnectAll();
+	if (isEngaged()) {
+		spdlog::error("MqttClient::EventLoopHook::engage() - Already engaged.");
+		return;
+	}
 
-	m_fdnRead = {};
-	m_fdnWrite = {};
-	m_miscTaskTimer = {};
-}
+	if (socket < 0) {
+		spdlog::error("MqttClient::EventLoopHook::engage() - Invalid socket.");
+		return;
+	}
 
-void MQTT::onFileDescriptorNotifierTriggeredRead()
-{
-	auto result = loop_read();
-	checkMosquitoppResultAndDoDebugPrints(result, "loop_read()");
+	readOpNotifier = std::make_unique<FileDescriptorNotifier>(socket, FileDescriptorNotifier::NotificationType::Read);
+	writeOpNotifier = std::make_unique<FileDescriptorNotifier>(socket, FileDescriptorNotifier::NotificationType::Write);
+
+	readOpNotifier->triggered.connect(&MqttClient::onReadOpRequested, parent);
+	writeOpNotifier->triggered.connect(&MqttClient::onWriteOpRequested, parent);
+	miscTaskTimer->timeout.connect(&MqttClient::onMiscTaskRequested, parent);
```

Comment:
Unfortunately we cannot move this to `init()` because `readOpNotifier` and `writeOpNotifier` need to have FDNs assigned to them when we setup the connections, this isn't the case in `init()`.

Reply:
I meant only the line:
```
miscTaskTimer->timeout.connect(&MqttClient::onMiscTaskRequested, parent);
```
This one merely sets up the connection so I guess it can happen once in `init()`. `engage`/`disengage` would then only start and stop the timer.

Reply:
Ah, right. Sure. Done.